### PR TITLE
Live: fast path for live push - cache API keys, bypass Macaron

### PR DESCRIPTION
**Pull request by FZambia**
**What this PR does / why we need it**:

We came across a fact that request latency is pretty big for our use case - one of our early adopters wants to conistently push data from Telegraf to Grafana with minimal possible latency. Telegraf can output data over HTTP at the moment.

Original latency for each request was 10-14ms, with removing middlewares and caching API key strategy (the biggest win here actually since there is a pretty expensive math behind - ~10ms overhead per request) we got 1-3ms, with further removing Macaron from request path we got 700-1000µs which seems acceptable for us at the moment. This also results into less generated garbage so we get smaller GC pauses (which is good since we get less occasional errors from Telegraf that its output endpoint can't answer fast enough).

What do you think about this draft? Can we proceed with it for push endpoint?

### CODE REVIEW

File: pkg/services/live/push/push.go
Comment (aknuds1) [inline pkg/services/live/push/push.go:32]: Nit, but can we avoid exporting this?
```suggestion
// cachedKey ...
type cachedKey struct {
```

File: pkg/services/live/push/push.go
Comment (aknuds1) [inline pkg/services/live/push/push.go:171]: `ioutil` is deprecated:
```suggestion
	body, err := io.ReadAll(req.Body)
```

**Review by aknuds1 — COMMENTED** [review]
AFAICT looks good.

File: pkg/services/live/push/push.go
Comment (FZambia) [reply pkg/services/live/push/push.go:32]: Yeah, definitely 👍

File: pkg/api/http_server.go
Comment (ryantxu) [inline pkg/api/http_server.go:117]: rather than checking on each request, should we only register this handler if live is enabled

Comment (FZambia) [reply pkg/api/http_server.go:117]: this will require different `http.Handler` implementations - which in turn makes code harder to follow.  The overhead should be really negligible since no allocations here. Also maybe `IsLiveEnabled` check will go away soon?

File: pkg/services/live/push/push.go
Comment (marefr) [inline pkg/services/live/push/push.go:57]: This will be logged even if it's not enabled so prefer debug.

```suggestion
	logger.Debug("Live Push Gateway initialization")
```

File: pkg/services/live/push/push.go
Comment (marefr) [inline pkg/services/live/push/push.go:71]: If you're not planning to do any background processing in here you can just skip implementing `Run`. See https://github.com/grafana/grafana/blob/438b403acc3e4a374872104a2ef1a510abdd5075/pkg/registry/registry.go#L123-L130

File: pkg/services/live/push/push.go
Comment (marefr) [inline pkg/services/live/push/push.go:93]: We already have similar code here: https://github.com/grafana/grafana/blob/438b403acc3e4a374872104a2ef1a510abdd5075/pkg/services/contexthandler/contexthandler.go#L140-L192

Possible to create something that can be reused by both to decrease code duplication?

Comment (marefr) [reply pkg/services/live/push/push.go:93]: Noticed the comment below now. Thinks that a good idea to do either now or slightly after this has been merged

File: pkg/services/live/push/push.go
Comment (marefr) [inline pkg/services/live/push/push.go:133]: Feels like this case will never happen or I might missing something. You probably need to store when api key expires in the cache to be able to invalidate it - otherwise Grafana would allow requests for a cached api key that in theory has expired.

**Review by marefr — CHANGES_REQUESTED** [review]
Looks good in general. Added some comments.

Perhaps the biggest concern is that by short-cutting macaron you wont get any macaron  middleware applied to your push requests. I understand you don't want all of that since you want to minimize the latency.

However, you should at least have [recovery](https://github.com/grafana/grafana/blob/master/pkg/middleware/recovery.go#L106) in place so a panic wont crash the grafana-server. 

Would also suggests the following, but up to you:
- [logger](https://github.com/grafana/grafana/blob/master/pkg/middleware/logger.go): convenient to be able to turn on router logging for all requests when investigating problems/issues. 
- [request metrics](https://github.com/grafana/grafana/blob/master/pkg/middleware/request_metrics.go) to be able to understand how many push requests (per channel?) Grafana receives and the latency of handling those.

File: pkg/services/live/push/push.go
Comment (FZambia) [reply pkg/services/live/push/push.go:133]: Yep - there is a comment about this, I have not implemented cache invalidation routine yet since wanted to get a green light with general idea about bypassing Macaron and caching keys.

File: pkg/services/live/push/push.go
Comment (FZambia) [reply pkg/services/live/push/push.go:93]: Yes, planning to do this inside this pr

File: pkg/services/live/push/push.go
Comment (FZambia) [reply pkg/services/live/push/push.go:71]: Thanks, maybe this should be useful for stopping cache invalidation routine.

Comment (marefr) [reply pkg/services/live/push/push.go:71]: Maybe. But maybe easier to handle it during the request?

Comment (FZambia) [reply pkg/services/live/push/push.go:71]: What to you mean? The main latency reduction here is removing token check on each request. I was planning to add a goroutine that will periodically invalidate/remove cached keys. Doing this during request will result into the same 10+ ms latency. Or you suggest sth else?

Comment (marefr) [reply pkg/services/live/push/push.go:71]: Related to my https://github.com/grafana/grafana/pull/33212#discussion_r617698530 where I suggest storing the time when the token expires in the cache and then doing the expiration check in the request shouldn't add any latency. But would you like to do it in the background instead that should also work. However might be a risk of more contention/locking of the mutex?

Comment (FZambia) [reply pkg/services/live/push/push.go:71]: Ah, right - I see now, we have a time of token expiration. Cool - of course better to just check expiration during a request then.

Comment (FZambia) [reply pkg/services/live/push/push.go:71]: Hmm, though can be revoked - since tokens long lived usually this can be a problem

Comment (FZambia) [reply pkg/services/live/push/push.go:71]: So I suppose we need to to both: invalidate periodically and check expiration on each request for existing token.

Comment (marefr) [reply pkg/services/live/push/push.go:71]: Yes you're right. Didn't think of that :+1: 

Another thing that I just came to think about is in the case you have many channels/streams data being pushed to, there's a chance of contention/locking. If that's a potential problem you might want to include the channel/stream in the cache key identifier. But maybe something to consider later.

Comment (FZambia) [reply pkg/services/live/push/push.go:71]: Also: looks like there could be tokens without expiration since `Expires` is nullable:

```go
type ApiKey struct {
        ...
	Expires *int64
}
```

Comment (FZambia) [reply pkg/services/live/push/push.go:71]: > If that's a potential problem you might want to include the channel/stream in the cache key identifier. But maybe something to consider later.

Yep, this is definitely an advice for a future

Comment (marefr) [reply pkg/services/live/push/push.go:71]: Yes, there could be tokens without expiration

**Comment by FZambia** [comment]
So after cleaning up and finishing invalidation: it works, endpoint responds fast enough.

The bad thing here is that while I added basic recovery here I think we can't simply add existing metrics and logger middleware - those are pretty Macaron-specific and pretty heavy.

@ryantxu what do you think if we switch back to your original idea with adding WebSocket output plugin for Telegraf? I just noticed that it already has [TCP/UDP socket write without waiting for any response back](https://github.com/influxdata/telegraf/blob/master/plugins/outputs/socket_writer/socket_writer.go#L155) - so in this perspective we don't have to invent async response awaiting for WS there too. Also Telegraf already has indirect dependency to Gorilla WebSocket in go.sum.

File: pkg/services/live/push/push.go
Comment (ryantxu) [reply pkg/services/live/push/push.go:68]: what about using a library like:
https://github.com/patrickmn/go-cache

This has builtin support to the periodic cleanup and checking individual entries when you get them

**Comment by FZambia** [comment]
@marefr @aknuds1 I updated pr to the form I believe is ready to review, but currently we also evaluating an option to add WebSocket output plugin to Telegraf (to avoid going through all middleware on each request but still have a possibility to utilize same HTTP port). WebSocket also has some downsides unfortunately  - it does not allow to control publication errors without implementing async protocol on top of WS connection. If WebSocket idea will work for us then we can maybe not proceed with changes here. Maybe you have any ideas regarding all of this.

**Comment by ryantxu** [comment]
@FZambia -- we are not moving forward with this PR correct?

**Comment by FZambia** [comment]
@ryantxu correct, closing.

**Pull request closed without merging**
